src/providers/aws.rs

```diff
@@ -0,0 +1,858 @@
+use std::io::Write;
+use std::sync::Arc;
+use std::{thread, time};
+
+struct UbuntuAmi(String);
+
+impl From<Region> for UbuntuAmi {
+    fn from(r: Region) -> Self {
+        // https://cloud-images.ubuntu.com/locator/
+        // ec2 20190814 releases
+        UbuntuAmi(
+            match r {
+                Region::ApEast1 => "ami-e0ff8491",               // Hong Kong
+                Region::ApNortheast1 => "ami-0cb1c8cab7f5249b6", // Tokyo
+                Region::ApNortheast2 => "ami-081626bfb3fbc9f49", // Seoul
+                Region::ApSouth1 => "ami-0cf8402efdb171312",     // Mumbai
+                Region::ApSoutheast1 => "ami-099d318f80eab7e94", // Singapore
+                Region::ApSoutheast2 => "ami-08a648fb5cc86fb74", // Sydney
+                Region::CaCentral1 => "ami-0bc1dd4eb012a451e",   // Canada
+                Region::EuCentral1 => "ami-0cdab515472ca0bac",   // Frankfurt
+                Region::EuNorth1 => "ami-c37bf0bd",              // Stockholm
+                Region::EuWest1 => "ami-01cca82393e531118",      // Ireland
+                Region::EuWest2 => "ami-0a7c91b6616d113b1",      // London
+                Region::EuWest3 => "ami-033e0056c336ecff0",      // Paris
+                Region::SaEast1 => "ami-094c359b4d8c6a8ca",      // Sao Paulo
+                Region::UsEast1 => "ami-064a0193585662d74",      // N Virginia
+                Region::UsEast2 => "ami-021b7b04f1ac696c2",      // Ohio
+                Region::UsWest1 => "ami-056d04da775d124d7",      // N California
+                Region::UsWest2 => "ami-09a3d8a7177216dcf",      // Oregon
+                x => panic!("Unsupported Region {:?}", x),
+            }
+            .into(),
+        )
+    }
+}
+
+impl Into<String> for UbuntuAmi {
+    fn into(self) -> String {
+        self.0
+    }
+}
+
+/// A descriptor for a particular machine setup in a tsunami.
+///
+/// The `setup` argument is called once for every spawned instances of this type with a handle
+/// to the target machine. Use [`Machine::ssh`](struct.Machine.html#structfield.ssh) to issue
+/// commands on the host in question.
+#[derive(Clone)]
+pub struct MachineSetup {
+    region: Region,
+    instance_type: String,
+    ami: String,
+    setup: Option<Arc<dyn Fn(&mut ssh::Session, &slog::Logger) -> Result<(), Error> + Send + Sync>>,
+}
+
+impl super::MachineSetup for MachineSetup {
+    type Region = String;
+
+    fn region(&self) -> Self::Region {
+        self.region.name().to_string()
+    }
+}
+
+impl PartialEq for MachineSetup {
+    fn eq(&self, other: &Self) -> bool {
+        self.ami == other.ami && self.instance_type == other.instance_type
+    }
+}
+
+impl Eq for MachineSetup {}
+
+impl std::hash::Hash for MachineSetup {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.instance_type.hash(state);
+        self.ami.hash(state);
+    }
+}
+
+impl Default for MachineSetup {
+    fn default() -> Self {
+        MachineSetup {
```

Comment:
These defaults should all be documented somewhere.

Reply:
In fact, does it even make sense to provide a `impl Default` for `MachineSetup`? It strikes me as weird. When would you ever not care?

Reply:
why not :smile:? These defaults are reasonable, and users can replace the ones they want.